infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/ReplicationHandlerMethods.java

```diff
@@ -0,0 +1,205 @@
+import org.corfudb.runtime.proto.service.CorfuMessage;
+import org.corfudb.runtime.proto.service.CorfuMessage.RequestPayloadMsg;
+import org.corfudb.runtime.proto.service.CorfuMessage.ResponsePayloadMsg;
+
+import javax.annotation.Nonnull;
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Slf4j
+public class ReplicationHandlerMethods {
+
+    private final Map<String, String> timerNameCache = new HashMap<>();
+
+    /**
+     * The request handler map.
+     */
+    private final Map<RequestPayloadMsg.PayloadCase, HandlerMethod> requestHandlerMap;
+
+    /**
+     * The response handler map.
+     */
+    private final Map<ResponsePayloadMsg.PayloadCase, HandlerMethod> responseHandlerMap;
+
+    /**
+     * A functional interface for clinet/server request/response handlers. The handlers should
+     * be fast and not block. If a handler blocks for an extended period of time, it will
+     * exhaust the server's thread pool. I/O and other long operations should be handled
+     * on another thread.
+     */
+    @FunctionalInterface
+    public interface HandlerMethod {
+        void handle(CorfuMessage.RequestMsg req,
+                    CorfuMessage.ResponseMsg res,
+                    @Nonnull IClientServerRouter r);
+    }
+
+    /**
+     * Construct a new instance of ReplicationHandlerMethods.
+     */
+    public ReplicationHandlerMethods() {
+        requestHandlerMap = new ConcurrentHashMap<>();
+        responseHandlerMap = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Given a {@link CorfuMessage.RequestMsg} return the corresponding handler.
+     *
+     * @param request   The request message to handle.
+     * @return          The appropriate {@link HandlerMethod}.
+     */
+    protected HandlerMethod getRequestHandler(CorfuMessage.RequestMsg request) {
+        return requestHandlerMap.get(request.getPayload().getPayloadCase());
+    }
+
+    /**
+     * Given a {@link CorfuMessage.ResponseMsg} return the corresponding handler.
+     *
+     * @param response   The respose message to handle.
+     * @return          The appropriate {@link HandlerMethod}.
+     */
+    protected HandlerMethod getResponseHandler(CorfuMessage.ResponseMsg response) {
+        return responseHandlerMap.get(response.getPayload().getPayloadCase());
+    }
+
+    /**
+     * Handle an incoming Corfu request message.
+     *
+     * @param req       The request message to handle.
+     * @param r         The server router.
+     */
+    @SuppressWarnings("unchecked")
+    public void handle(CorfuMessage.RequestMsg req, CorfuMessage.ResponseMsg res, IClientServerRouter r) {
```

Comment:
what is the difference between this method and the one on L47?

Reply:
this one is invoking the handler method in LogReplicationServer.
And the other one is a functional interface, that maps to the handler methods which is stored in memory and the router uses this in memory structure and calls this handle().